docs/proposals/2018-05-declarative-delete-field.md

```diff
@@ -0,0 +1,83 @@
+# Declarative Delete Field
+
+Concept committed to [during 2018 April 12F2F](https://docs.google.com/document/d/1O7_fws7hwZ6qV3okAjbV5qdFaEq4bt1LYJ9bTu-IwvU)
+
+This document 2018-05.
+
+## Abstract
+
+Add a field to instance and binding to indicate that they have been asked to
+send a DELETE to the backend.
+
+## Motivation
+
+When the Originating-Identity header is used, the broker and refuse a delete
+call when the user info in the Originating-Identity header does not have
+permissions to delete the instance or binding. As this happens after kubernetes
+has accepted the DELETE as an true indication to do a delete, it has set the
+DeletionTimestamp, which is an irreversible step on the way to carrying out the
+necessary actions to delete a kubernetes resource.
+
+## Proposed Design
+
+ - Add a field to indicate deletion
+ - Add a condition to indicate permanent end state of deletion success.
+
+Once the field is activated the controller deprovisions the broker
+resource. When successfully deleted the controller then deletes the
+kubernetes resource.
+
+A feature flag to enable/disable this behavior.
+
+RBAC changes to allow the controller to issue deletes.
+
+Helm chart changes to enable the flag and the rbac rules.
+
+### API Resource Changes
+
+Add a condition for ServiceInstanceConditionType, ServiceBindingConditionType.
+```go
+  ServiceInstanceConditionDeleted ServiceBindingConditionType = "Deleted"
+  ServiceBindingConditionDeleted ServiceBindingConditionType = "Deleted"
+```
+
+Add a field to spec of ServiceInstance, ServiceBinding.
+```go
+  ExistsExternally bool `json:"existsExternally,omitempty"`
+```
+
+This field is set to true by default. The user indicates that it wants
+to delete the backing resource by setting this field to false. The
+controller will see this update and attempt to delete the backing
+broker resource. If the broker delete succeeds, it sets the deleted
+status condition.
+
+If the field is set to false, no updates are allowed besides
```

Comment:
Why this limitation? Perhaps you want to stage a bunch of updates to a non-existent resource.

Reply:
It is a simpler case.

What would it mean to update something that is going away? If there is some update required for the deletion to go through, I would make the update first and make sure it is committed before deleting.

What would staging updates to a non-existent resource mean? What would it do? Once the resource is gone, it is not going to come back.

Reply:
ExistsExternally=false means the broker holds no resource for you and you are staging changes or you shutdown the external resource. flipping to true means provision it again. 

> What would it mean to update something that is going away?

I am staging changes, or I am getting a whole set of resources preped for a deploy. Or I want to shut the services down for a bit and then intend to bring them back later.

> before deleting.

Stop thinking of it as deleting, it is deprovisiong, and it can be provisioned again in the same way.

> Once the resource is gone, it is not going to come back.

Let it come back if ExistsExternally = true.

Reply:
https://github.com/kubernetes-incubator/service-catalog/pull/2037/files#r189993692
"flipping to true means provision it again." That is not the workflow I was suggesting to implement or support.
"getting a whole set of resources preped for a deploy. Or I want to shut the services down for a bit and then intend to bring them back later." not the workflow suggested or intended to implement or support.
" it is deprovisiong, and it can be provisioned again in the same way." I'd prefer to have only one way to do things. 
"Let it come back if ExistsExternally = true." not the workflow suggested or intended to implement or support.